src/Stateless/StateConfiguration.cs

```diff
@@ -508,6 +508,26 @@ public StateConfiguration OnExit(Action<Transition> exitAction, string exitActio
             /// <returns>The receiver.</returns>
             public StateConfiguration SubstateOf(TState superstate)
             {
+                var State = _representation.UnderlyingState;
+
+                // Check for accidental identical cyclic configuration 
+                if (State.Equals(superstate)) { throw new ArgumentException($"Configuring {State} as a substate of {superstate} creates an illegal cyclic configuration."); }
+
+                // Check for accidental identical nested cyclic configuration 
+                List<TState> superstates = new List<TState>();
+                superstates.Add(State);
+
+                // Build list of super states 
+                var activeRepresentation = _lookup(superstate);
+                while (activeRepresentation.Superstate != null)
+                {
+                    superstates.Add(activeRepresentation.Superstate.UnderlyingState);
```

Comment:
Using `HashSet` and moving the check here we can get away from the `GroupBy` below.

```csharp
if (superstates.Contains(activeRepresentation.Superstate.UnderlyingState))
    throw new ArgumentException($"Configuring {State} as a substate of {superstate} creates an illegal nested cyclic configuration.");
```